tests/nightly/test_large_array.py

```diff
@@ -279,6 +279,19 @@ def test_diag():
     assert_almost_equal(r.asnumpy(), np.diag(a_np, k=k))
 
 
+def test_ravel_and_unravel():
+    idxs_2d = [[LARGE_X-1,LARGE_X-100,6],[SMALL_Y-1,SMALL_Y-10,1]]
```

Comment:
Please use C++ with mocking of the memory allocation instead of actually allocating memory.

Although this is nightly, I don't think it's necessary for these kind of tests to actually allocate real memory.

Reply:
@perdasilva

Reply:
We can add C++ unit test to mock the memory allocation. However, I thought the nightly test is  used for end-to-end testing purpose to mimic the real deployment situation, or is it not?

Reply:
They are to some extent, but in same cases the cost vs return ratio is not that good. While smoke tests are certainly necessary in some cases, resource intensive ones on the other hand can usually be abstracted away. If we assume that memory access on the system side is properly working and only mock this very laster layer, this result should be pretty much as meaningful as running it end to end.

Reply:
@marcoabreu : I didn't understand your last comment. What did you mean by "very laster layer" ? All these tests are single operator tests. This tests I combined because its easier to test when combined together

Reply:
because ravel and unravel were reverse of each other

Reply:
@marcoabreu ??

Reply:
When I'm talking about layers in this case, I'm talking about the physical execution on a machine and not about machine learning operator layers.

Specifically in this case I'm talking about mocking memory accesses. Read and writes would access a virtual array array that doesn't have the full capacity on physical memory but instead consists of procedually generated data.

In various languages there's the concept of Iterators https://jeffknupp.com/blog/2013/04/07/improve-your-python-yield-and-generators-explained/ where a function can be invoked that will deliver a stream of entries. If you would call ```.toList()``` that would then create a physically mapped array (which we wouldn't want). Instead, you could call ```.get(1243952)``` which would iterate the array but not allocate for the full size but only for that particular entry. So support writes, you could have custom lookups that would allow setting certain values in a sparse fashion. The iterator function could be something like
```
float64 getValue(int64 index) 
{
    dictValue = _internalDict.get(index);
    if (dictValue != null)
        return dictValue;
    else
        return index; //Replace with generator function 
}

setValue(int64 index, float64 value) 
{
    _internalDict.set(index, value)
}
```

With this concept you'd be able to test infinitely big arrays as long as you only use sparse data to test (the size of the dict would depend on the number of specifically set entries). Since the purpose of these large array tests is not to test dense data (it wouldn't make a difference between sparse and dense besides performance I assume), this type of generator model would allow us to execute these tests in an efficient way.

The only thing we'd have to mock at this point are the getters and setters that try to read/write from/to physical memory. The calls would be redirected to the pseudo-code above.


I'm aware that we do not support this type of behaviour out of the box, but I think that we're getting to a point where we should take a step back and reconsider the approaches we would like to do in such resource-heavy cases. Of course, we wouldn't be testing the physical memory access with my method, but I think that we can assume that memory accesses are properly working or otherwise we'd notice in quite a few other cases.

Reply:
Sorry for my belated response. I have copied my comment to dev@ to bring this topic to a wider audience.

Reply:
Hi @marcoabreu, thanks a lot for your suggestion and it's a very good learning for me too. The large tensor support is crucial to DGL (Deep Graph Library) that is using MXNet and planning to release with support of large tensors. Without this test, we cannot make this feature production ready.

While I appreciate the direction of testing you suggested and definitely would like to explore it further, I am also afraid that not adding tests before we implement the approach you suggested will block the release of DGL and other MXNet projects that depend on large tensors as well.

Can we have a seperate project to address the test efficiency (we may need some expertise from CI team for this project) while moving on with the large tensor project? Please let me know if you have other concerns or advice. Thank you

Reply:
@marcoabreu : We have to test cases where we have to allocate large tensors on actual physical memory. That is the whole purpose of adding these tests. Without which we cannot support or even guarantee large dense tensor support on MXNet. 

The purpose of these tests are not at all to test only Sparse. Currently we are not addressing performance issues with large dense arrays but trying to support them. From your reply it seems like you are under the impression that everything should work fine if tensor size >2^32 elements(Please correct me if I am wrong). But that's not the case and the operator simply segfaults or gives unpredictable behavior. That is why is necessary to test them with such large arrays/tensors.

Reply:
@marcoabreu : Let me know if you have other unanswered queries that will help you understand the use case better and unblock this PR.

Reply:
Hey,

I understand the importance of large tensors and I'm really looking forward to them as well. Don't get me wrong, I'm not underestimating the work that comes with supporting them.

We already had a history of issues that were caused by too large allocations. Thus, for the sake of keeping CI stable (independent on whether we're running on per-PR or nightly) and to adhere to a high test standard, I'd prefer if we would not delay adressing the test strategy.

I think there's a misunderstanding. With my proposed approach, you would still be able to make a full test of a single operator. But instead of randomly generating the input tensor, you would use a generator model that gives you predictable and procedually generatable input tensor instead of a large and randomly generated one.

The idea here would then be to give this procedually generated input tensor to the operator and since it's a predicutable input, you can also have a predictable output that you can verify. On the way, the operator will run as usual, it's just that the source of the data is not physical memory but generated values. 

The only case that wouldn't be supported is the chaining of operators since that would require storing the intermediary values. 

The idea here is that (in whatever fashion you'd like) we generate fake input and output tensors that return predictable and valid data, but doesn't actually consume memory on the RAM. From MXNets perspective, it shouldn't matter where the real values is coming from. For the sake of an operator, it's important that it actually retrieves proper values and that would be the case here.

Reply:
@marcoabreu  I still don't understand how will I know that on actual large tensor after memory allocation would it work as expected or not. Procedurally generating the values will indeed give me random values on random locations that are indexed at locations > 2^32. But the input should come from RAM. it does matter. It needs to be tested on actual physical memory so there are no surprises. If nighly tests are not feasible do we have a weekly test where we can run this ?